addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/UserControlledOpenRedirectScanner.java

```diff
@@ -119,7 +132,14 @@ private void checkUserControllableLocationHeaderValue(
                     || paramValue.equalsIgnoreCase(token)
                     || (responseLocation.indexOf("://") > 0
                             && paramValue.indexOf(responseLocation) >= 0)) {
-                raiseAlert(msg, id, param.getName(), paramValue, responseLocation);
+                // requestDomain is null if some exception occurred. if requestDomain is equal to
+                // location header then it is not considered as an issue.
+                // https://github.com/zaproxy/zaproxy/issues/5289
+                if (token != null
+                        || requestDomain == null
+                        || !requestDomain.equalsIgnoreCase(domain)) {
```

Comment:
Are you totally sure about this logic? Ex: If token isn't null then it's a local path, which implies same domain. But this returns an alert and fails:
```java
@Test
public void foobar() {
    // Given
    HttpMessage msg = createMessage();
    TreeSet<HtmlParameter> params = new TreeSet<HtmlParameter>();
    params.add(new HtmlParameter(HtmlParameter.Type.url, "place", "/images"));
    msg.setGetParams(params);
    msg.getResponseHeader().setStatusCode(HttpStatusCode.MOVED_PERMANENTLY);
    msg.getResponseHeader().setHeader(HttpHeader.LOCATION, "/images");
    // When
    rule.scanHttpResponseReceive(msg, -1, createSource(msg));
    // Then
    assertThat(alertsRaised.size(), equalTo(0));
}
```

I think should should simply be, the following?
```java
if (requestDomain == null 
        || !requestDomain.equalsIgnoreCase(domain))
```

Reply:
My queries @kingthorin , 
1) in that case (Pointed out by you) we can remove the token check from our scanner, as it is not required because it is a relative path.
2) in case attacker removes SSLie redirects user to http from https, do we need to take this into account for redirect scan?
3) what should be the behavior when ParamValue is say "http://evil.com/abc" and location header is say "http://evil.com/xyz" and request Url is "http://example.com" ?
eg testcase -
```
@Test
    public void
            checkingLogic() {
        // Given
        HttpMessage msg = createMessage();
        TreeSet<HtmlParameter> params = new TreeSet<HtmlParameter>();
        params.add(new HtmlParameter(HtmlParameter.Type.url, "place", "http://ex.com/abc"));
        msg.setGetParams(params);
        msg.getResponseHeader().setStatusCode(HttpStatusCode.MOVED_PERMANENTLY);
        msg.getResponseHeader().setHeader(HttpHeader.LOCATION, "http://ex.com/xyz");
        // When
        rule.scanHttpResponseReceive(msg, -1, createSource(msg));
        // Then
        assertThat(alertsRaised.size(), equalTo(0));
    }
```

Also there are some issues with the code logic today i found.
Code lines -
```
if (paramValue.equalsIgnoreCase(protocol)
                    || paramValue.equalsIgnoreCase(domain)
                    || paramValue.equalsIgnoreCase(token)
                    || (responseLocation.indexOf("://") > 0
                            && paramValue.indexOf(responseLocation) >= 0))
```
here as per above "if" clause even if protocol matches between param value and location header we are raising alert.

Test Case depicting same -
```
 @Test
    public void
            checkingLogic() {
        // Given
        HttpMessage msg = createMessage();
        TreeSet<HtmlParameter> params = new TreeSet<HtmlParameter>();
        params.add(new HtmlParameter(HtmlParameter.Type.url, "place", "http"));
        msg.setGetParams(params);
        msg.getResponseHeader().setStatusCode(HttpStatusCode.MOVED_PERMANENTLY);
        msg.getResponseHeader().setHeader(HttpHeader.LOCATION, "http://ex.com/sadasd");
        // When
        rule.scanHttpResponseReceive(msg, -1, createSource(msg));
        // Then
        assertThat(alertsRaised.size(), equalTo(1));
    }
```

Reply:
I'll be honest the new set of ORs just seemed strange to me so I dug into them a bit, I didn't dig into the existing functionality.

Reply:
oh ok. I have done the changes as suggested by you and also some other changes. please review.

Reply:
> 2. in case attacker removes SSLie redirects user to http from https, do we need to take this into account for redirect scan?

I don't think that should matter. https://example.com and https://example.com should be treated the same from my point of view.

> 3. what should be the behavior when ParamValue is say "http://evil.com/abc" and location header is say "http://evil.com/xyz" and request Url is "http://example.com" ?
eg testcase -

I think it this case we shouldn't alert, because it "seems" like some processing was done by the app. (If both ParamValue and Location are "http://evil.com/abc" then it should, which seems to be what currently happens [in 25d60a4].)

Reply:
hmm.

Reply:
Keep in mind this is a passive scanner not active.

Reply:
yes yes, got your point.

thanks